### PR TITLE
feat(Icon): composition

### PR DESCRIPTION
**Pull request by frassinier**
**What is the problem this PR is trying to solve?**
The icon can only be instantiated by its name for Button, Link or InlineMessage

**What is the chosen solution to this problem?**
Now, the icon could be the React element

**Please check if the PR fulfills these requirements**

- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: src/components/Button/Button.tsx [hunk 1]
```diff
@@ -2,14 +2,14 @@ import React from 'react';
...
+import { Icon, IconProps } from '../Icon/Icon';
 import Loading from '../Loading';
 
 import * as S from './Button.style';
 
 export type ButtonProps = StyledProps<any> & {
-	/** The icon name of the button */
-	icon?: IconName;
+	/** The icon of the button */
+	icon?: IconName | IconProps;
```
Comment (jmfrancois) [inline src/components/Button/Button.tsx:12]: props type is not aligned with the body

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -30,8 +30,17 @@ const Button: React.FC<ButtonProps> = React.forwardRef(
...
-			{loading && <Loading className="btn__loading" name={icon} aria-hidden />}
-			{!loading && icon && <Icon className="btn__icon" name={icon} />}
+			{loading && <Loading className="btn__loading btn__icon" name={icon} aria-hidden />}
+			{!loading &&
+				icon &&
+				(typeof icon === 'string' ? (
+					<Icon className="btn__icon" name={icon} />
+				) : (
+					React.cloneElement(icon, {
+						...icon.props,
```
Comment (jsomsanith-tlnd) [inline src/components/Button/Button.tsx:40]: You can remove that, `React.cloneElement` already merges the element's props with the ones you pass.

File: src/components/Icon/Icon.tsx [hunk 3]
```diff
@@ -159,41 +159,44 @@ export const Icon = React.forwardRef<SVGSVGElement, IconProps>(
...
+			[`tc-icon-name-${name}`]: !(isImg || isRemote),
+		});
+
 		if (isImg) {
 			return (
-				<img alt="" src={name.substring(4)} className="tc-icon" {...accessibility} {...rest} />
+				<img
+					alt=""
+					src={name.substring(4)}
+					className={classnames('tc-icon', classname)}
```
Comment (jsomsanith-tlnd) [inline src/components/Icon/Icon.tsx:171]:
```suggestion
					className={classname}
```

`tc-icon` is already in `classname`

File: src/components/Icon/Icon.tsx [hunk 4]
```diff
@@ -159,41 +159,44 @@ export const Icon = React.forwardRef<SVGSVGElement, IconProps>(
...
-
 		if (isRemote && content && !isRemoteSVG) {
-			const classNames = classnames('tc-icon', className);
-			iconElement = (
+			return (
 				<img
 					alt=""
 					src={name.replace('remote-', '')}
-					className={classNames}
+					className={classnames('tc-icon', className)}
```
Comment (jsomsanith-tlnd) [inline src/components/Icon/Icon.tsx:183]:
```suggestion
					className={className}
```

Same here

File: src/components/Link/Link.tsx [hunk 5]
```diff
@@ -60,12 +60,28 @@ const Link: React.FC<LinkProps> = React.forwardRef(
 				ariaDisabled={disabled ? 'true' : null}
 				ref={ref}
 			>
-				{iconBefore && <Icon className="link__icon link__icon--before" name={iconBefore} />}
+				{iconBefore &&
+					(typeof iconBefore === 'string' ? (
+						<Icon className="link__icon link__icon--before" name={iconBefore} />
+					) : (
+						React.cloneElement(iconBefore, {
+							...iconBefore.props,
```
Comment (jsomsanith-tlnd) [inline src/components/Link/Link.tsx:68]: You can remove that

File: src/components/Link/Link.tsx [hunk 6]
```diff
@@ -60,12 +60,28 @@ const Link: React.FC<LinkProps> = React.forwardRef(
...
 				{isExternal && !hideExternalIcon && (
 					<Icon className="link__icon link__icon--external" name="talend-link" />
 				)}
-				{iconAfter && <Icon className="link__icon link__icon--after" name={iconAfter} />}
+				{iconAfter &&
+					(typeof iconBefore === 'string' ? (
+						<Icon className="link__icon link__icon--after" name={iconAfter} />
+					) : (
+						React.cloneElement(iconAfter, {
+							...iconAfter.props,
```
Comment (jsomsanith-tlnd) [inline src/components/Link/Link.tsx:81]: You can remove that

**Review by jsomsanith-tlnd — APPROVED** [review]

**Pull request merged**